src/lang/eval/EvalUtil.ml

```diff
@@ -175,126 +167,49 @@ module Configuration = struct
-      | Some x -> pure @@ (snd x, G_Load(snd x))
-      | None -> fail1 (sprintf "No field \"%s\" in contract.\n" i)
-            (ER.get_loc (get_rep k))
-
-  let map_update st m klist vopt =
-    let s = st.fields in
-    match List.find s ~f:(fun (z, _) -> z = (get_id m)) with
-    | Some (_, Map((_,vt), mlit)) ->
-      let rec recurser mlit' klist' vt' =
-        (match klist' with
-         (* we're at the last key, update literal. *)
-         | [k] -> 
-            (match vopt with
-            | Some v ->
-              Caml.Hashtbl.replace mlit' k v;
-              pure @@ (st, G_MapUpdate ((List.length klist), (Some v)))
-            | None ->
-              Caml.Hashtbl.remove mlit' k;
-              pure @@ (st, G_MapUpdate ((List.length klist), None)))
-         | k :: krest ->
-            (* we have more nested maps *)
-            (match Caml.Hashtbl.find_opt mlit' k with
-             | Some (Map((_, vt''), mlit'')) -> recurser mlit'' krest vt''
-             | None ->
-                if (is_some vopt) then (* not a delete operation. *)
-                  (* We have more keys remaining, but no entry for "k".
-                    So create an empty map for "k" and then proceed. *)
-                  let mlit'' = Caml.Hashtbl.create 4 in
-                  let%bind (kt'', vt'') = 
-                    (match vt' with
-                    | MapType (keytype, valtype) -> pure (keytype, valtype)
-                    | _ -> fail1 (sprintf "Cannot index into map %s due to non-map type" (get_id m))
-                            (ER.get_loc (get_rep m))
-                    )
-                    in
-                      Caml.Hashtbl.replace mlit' k (Map((kt'', vt''), mlit''));
-                      recurser mlit'' krest vt''
-                else
-                  (* No point removing a key that doesn't exist. *)
-                  pure @@ (st, G_MapUpdate ((List.length klist), None))
-              (* The remaining keys cannot be used for indexing as
-                 we ran out of nested maps. *)
-             | _ -> fail1 (sprintf "Cannot index into map %s. Too many index keys." (get_id m))
-                      (ER.get_loc (get_rep m))
-            )
-          (* this cannot occur. *)
-          | [] -> fail1 (sprintf "Internal error in updating map %s." (get_id m))
-                    (ER.get_loc (get_rep m))
-        )
-      in
-        recurser mlit klist vt
-    | _ -> fail1 (sprintf "No map field \"%s\" in contract.\n" (get_id m))
-            (ER.get_loc (get_rep m))
+      let%bind fval = fromR @@ StateService.fetch ~fname:k ~keys:[] in
+      match fval with
+      | (Some v, g) -> pure (v, g)
+      | _ -> fail1 (Printf.sprintf "Error loading field %s" i) (ER.get_loc (get_rep k))
+
+  let map_update m klist vopt =
+    match vopt with
+    | Some v -> fromR @@ StateService.update ~fname:m ~keys:klist ~value:v
+    | None -> fromR @@  StateService.remove ~fname:m ~keys:klist
```

Comment:
Same as l160 -- oh, I see, it all is now simply moved to `StateService.ml`. Now another question -- is it conceptually better or it's just moving the same logic around?

Reply:
It's not conceptually better. It is just moving the logic around. But that's because, in `StateService.ml`, we can now (in the near future) add code that can communicate with the blockchain / database (or whatever) and provide the state values on query. It doesn't have to be "in-memory" (although that mode will be supported too, for testing convenience)